spinoso-string/src/enc/utf8/mod.rs

```diff
@@ -1216,4 +1243,40 @@ mod tests {
         assert_eq!(s.get_char_slice(10..5), None);
         assert_eq!(s.get_char_slice(10..2), None);
     }
+
+    #[test]
+    fn index_with_default_offset() {
+        let s = Utf8String::from("f💎oo");
+        assert_eq!(s.index("f".as_bytes(), 0), Some(0));
+        assert_eq!(s.index("o".as_bytes(), 0), Some(2));
+        assert_eq!(s.index("oo".as_bytes(), 0), Some(2));
+        assert_eq!(s.index("ooo".as_bytes(), 0), None);
+    }
+
+    #[test]
+    fn index_with_different_offset() {
+        let s = Utf8String::from("f💎oo");
+        assert_eq!(s.index("o".as_bytes(), 1), Some(2));
```

Comment:
```console
[3.1.2] > "💎".unpack "C*"
=> [240, 159, 146, 142]
```

I suspect adding some test cases with needles as various subslices of the UTF-8 encoding for `💎` will reveal the bugs in relying on `ByteSlice::chars` above.

Reply:
Thanks! Indeed, I can reproduce the issue in 751be01.